pallets/parachain-staking/src/lib.rs

```diff
@@ -1650,6 +1658,56 @@ pub mod pallet {
 			<SelectedCandidates<T>>::put(collators);
 			(collator_count, delegation_count, total)
 		}
+
+		/// Apply the delegator intent for revoke and decrease in order to build the
+		/// effective list of delegators with their intended bond amount.
+		///
+		/// This will:
+		/// - if [DelegationChange::Revoke] is outstanding, set the bond amount to 0.
+		/// - if [DelegationChange::Decrease] is outstanding, subtract the bond by specified amount.
+		/// - else, do nothing
+		///
+		/// The intended bond amounts will be used while calculating rewards.
+		fn get_rewardable_delegators(
+			collator: &T::AccountId,
+		) -> Vec<Bond<T::AccountId, BalanceOf<T>>> {
+			<TopDelegations<T>>::get(collator)
+				.expect("all members of CandidateQ must be candidates")
+				.delegations
+				.into_iter()
+				.map(|mut bond| {
+					bond.amount = match Self::delegator_scheduled_requests(&bond.owner, &collator) {
```

Comment:
To make sure we're on the same page, this is the problem WRT performance. For every single top delegation we do a DB read here to see if there is a pending request, which is O(n), and this is all done in a single block when we change rounds.

I'm not quite clear on the performance of a DB lookup miss (e.g. we look something up in a storage map and get a `None`), but that wouldn't matter for worst case where every single delegator has a pending request.

Reply:
Yeah the second draft approach was known to have the same runtime/rw complexity and only reduce the PoV size by having smaller units. In general, I'd have loved to have some prometheus metrics on what are the usual patterns we could see around cancellation calls, average delegator count, etc. - but in absence of it and using the worst case scenario:

#### Option 1 - Delegator Indexed `DelegatorScheduledRequests`

**At each round we read the storage `d` delegators, per collator**

* When creating `CollatorSnapshot` we do `O(d)` reads. PoV is smaller in size compared to reading complete `DelegatorState` (`32` instead of `112` - see PR description).
* At each revoke, bond_decrease, cancel and execute request, the read/writes stay mostly same to "light" storages (size < `32`) and in some places 1 extra read for `DelegatorState`.

#### Option 2 -  Delegator Indexed `DelegatorScheduledRequests` + Collator Indexed `CollatorDelegatorScheduledRequests`
The inverted index storage can be a new one, since we never read `CollatorState` during rewards, only the `CollatorSnapshot`. Thus giving us no "free" read.

* At each revoke, bond_decrease, cancel and execute request we perform same read/writes as `Option 1` and 1 **extra** write - to write to the inverted index.
* When creating `CollatorSnapshot` we have a O(1) access to all pending requests.

_Note:_ This still depends on how often we get these extrinsics - if every delegator submits a request per Round either way, then both options converge (Option 2 being still slightly more expensive due to additional writes). We get `O(d)` reads to `DelegatorScheduledRequests` plus `O(d)` extra writes to `CollatorDelegatorScheduledRequests` and any additional remove operations.

`Option 2` also involves data duplication that needs to be maintained.

However, talking to @notlesh there was also the opinion that it's fine if we make extrinsics "heavier" (more r/w) since the users would pay the fee, instead of us "paying" for it each round.

I believe this could be a good discussion to be had on what we should pay attention to.

/cc @crystalin @librelois @4meta5